images/win/scripts/Installers/Validate-ToolCache.ps1

```diff
@@ -17,85 +17,114 @@ function Get-ToolcachePackages {
-
-                    if ($foundVersion -ne $null){
-
-                        $architectures = GetChildFolders -Path "$env:AGENT_TOOLSDIRECTORY\$SoftwareName\$foundVersion"
-
-                        $softwareArch = $softwarePackage.Arch
-
-                        if ($architectures -Contains $softwareArch) {
-                            $path = "$env:AGENT_TOOLSDIRECTORY\$SoftwareName\$foundVersion\$softwareArch"
-                            foreach ($test in $ExecTests)
-                            {
-                                if (Test-Path "$path\$test")
-                                {
-                                    Write-Host "$SoftwareName($test) $foundVersion($softwareArch) is successfully installed:"
-                                    Write-Host (& "$path\$test" --version)
-                                }
-                                else
-                                {
-                                    Write-Host "$SoftwareName($test) $foundVersion ($softwareArch) is not installed"
-                                    exit 1
-                                }
-                            }
-                            $description += "_Version:_ $foundVersion ($softwareArch)<br/>"
-                        }
-                        else
-                        {
-                            Write-Host "$env:AGENT_TOOLSDIRECTORY\$SoftwareName\$foundVersion does not include required architecture"
-                            exit 1
-                        }
-                    }
-                    else
-                    {
-                        Write-Host "$env:AGENT_TOOLSDIRECTORY\$SoftwareName\$version.* was not found"
-                        exit 1
-                    }
-                }
-
-                Add-SoftwareDetailsToMarkdown -SoftwareName $SoftwareName -DescriptionMarkdown $description
-            }
+        if (Test-Path "$Path\$test")
+        {
+            Write-Host "$SoftwareName($test) $SoftwareVersion($SoftwareArchitecture) is successfully installed:"
+            Write-Host (& "$Path\$test" --version)
         }
         else
         {
-            Write-Host "$env:AGENT_TOOLSDIRECTORY\$SoftwareName does not include any folders"
+            Write-Host "$SoftwareName($test) $SoftwareVersion($SoftwareArchitecture) is not installed"
             exit 1
         }
     }
-    else
+}
+
+function UpdateMarkdownDescription {
+    param (
+        [string]$Description,
+        [Parameter(Mandatory = $True)]
+        [string]$SoftwareVersion,
+        [Parameter(Mandatory = $True)]
+        [string]$SoftwareArchitecture
+    )
+    return $Description += "_Version:_ $SoftwareVersion ($SoftwareArchitecture)<br/>"
+}
+
+function ToolcacheTest {
+    param (
+        [Parameter(Mandatory = $True)]
+        [string]$SoftwareName,
+        [Parameter(Mandatory = $True)]
+        [string[]]$ExecTests
+    )
+
+    $markdownDescription = ""
```

Comment:
Can we declare it closer the place where it's used? Also, could you please confirm that we do not need to reset it in the cycle? (probably we need to declare it in the cycle itself)

Reply:
It is my fault. Thanks for you point! I moved `Add-SoftwareDetailsToMarkdown` function from the loop. In case when this function was in loop, the software details placed to Markdown several times for one software package